add sendToCosmos nonce

jtremback opened the pull request:
This tracks the work on this issue: https://github.com/althea-net/peggy/issues/48

- [x] Add logic to solidity to emit and increment nonce

- [x] Add logic to rust to claim with nonce

- [x] Add logic to Go to reject duplicate nonces

- [x] Get duplicate ERC20 deposits to Ethereum working

- [x] Maintain existing valset functionality

- No transaction batches yet
- No batch signature submissions yet.

jtremback commented:
Added a WIP for the rust code in "WIP rust code". It's not tested and I'd be very surprised if it works.

jkilpatr commented:
You can run the integration tests to see if you're rust code works, simply print the parsed nonce value. I think you also need to modify the code where we submit the actual SendToEth call to include a nonce.

jkilpatr commented on module/x/peggy/keeper/attestation.go:
```diff
@@ -32,12 +76,15 @@ func (k Keeper) AddClaim(ctx sdk.Context, claimType types.ClaimType, nonce types
 	// or false that are non-obvious. Great way to have a double-spend bug TODO refactor
+	// Jehan's note: This guards acceptance of a given claim. If this conditional does not return,
+	// tokens are moved on the next line.
 	if att.Certainty != types.CertaintyObserved || att.Status != types.ProcessStatusInit {
```
This conditional should be flipped. Easier to read. 

```
if att.Certainty == types.CertaintyObserved && att.Status == types.ProcessStatusInit {
current rest of the function
} else {
current if statmement
```

jkilpatr commented on module/x/peggy/keeper/attestation.go:
```diff
@@ -18,12 +18,56 @@ import (
+// comes in. So given that this function is common to both types of claims, it is a bad place to implement any batch logic.
+// We will have to think about how to split the code. We may want to make two versions of this whole function.
+//
+// Let's say there are two batches in the batch pool. Batches with nonce 101 and 102. Both are successfully submitted to the
```
this can be prevented by having the eth_signer enforce ordering. This is already a trusted role so I don't see too much of a difference adding this responsibility.

jtremback replied:
We covered this in chat earlier, but that seems very brittle and non obvious

jtremback commented:
This pull request has ballooned in complexity since adding a correctly functioning event nonce to be used to relay events from Ethereum also broke all the other functionality which was overloaded into the attestation mechanism.

I have greatly simplified how the attestation mechanism works, making it more well-suited to handling only the secure relaying of events from Eth into Cosmos by the validators. Unfortunately, this has also broken signature gathering for relaying batches and valset updates from Cosmos to Eth. But these needed to be separated anyway for clarity.

Summary:
- Greatly reduced the possible number of states that an attestation can be in. Previously an attestation could be in a total of 36 different states as a combination of the Certainty, ProcessStatus, and ProcessResult enums (4x3x3). The vast majority of these states were invalid by definition, and were not handled by the code. In fact, there are only two valid states for an attestation: when it does not have enough votes, it is not Observed, and it is not applied to the consensus state. Once it is Observed, it follows that it has been applied to the consensus state. Anything else represents a failure of consensus. Here is the new lifecycle for attestations:
    - From the time the first Claim about the Attestation is submitted, Observed is false
    - Once an attestation has enough votes, Observed is set to true, and the attestation is applied to the consensus state (e.g. moves tokens)
    - Attestations with Observed=true hang around indefinitely for now. They will be used by our slashing code to slash validators who do not vote. Once we write the slashing code, we will probably also figure out when to delete attestations.
- Added event nonce check as soon as a claim from a validator comes in. This allows them to fail fast and for that validator's eth signer to potentially recover before it becomes a consensus issue.
- Changed the vote-counting mechanism from a sum of validator powers, measured at the time the validator cast their vote, to an array of validators, whose powers are summed up each time we check whether an Attestation is Observed. This seems more robust to weird effects resulting from validation power changing while an Attestation is not Observed, but it still needs to be thought through again.
    - This votes array is also a lot to store, since it will be around 100 validator addresses. It may be possible to reduce this storage load by using a system similar to the contract, where we just store a digest of all the votes, and rely on an external node with access to the Cosmos events to submit proofs about the votes. For now we will keep things simple.
- Some minor refactors around how the last attestation nonce is stored. Previously, there was a bizarre gas optimization involving storing the nonce only in the key of the kv store. I don't understand it so I've temporarily replaced it by just storing a nonce in the normal way. We can add it back in as part of an optimization spike later.
- It appeared that validators stored and retrieved attestations by their nonce only. This would have resulted in validators only voting that an attestation with a given nonce had once been submitted, not voting on the actual content of the attestation. Perhaps there was something else preventing this from being the case? Anyway, I have replaced it with with a simple mechanism where attestations are stored under a concatenation of their eventNonce and the hash of their Details. This way, validators vote on the exact content of the attestation. There is likely some scope for optimization here, since it may be that all that actually needs to be stored about an attestation is its votes array and its Observed status.
 - Removed code having to do with MultiSigUpdates and bridge bootstrapping. It was decided not to rely on Cosmos having knowledge of the current multisig set, and we will probably ship in the short term with the bootstrapping procedure which was demonstrated during BrokenNet, which is secure, if clumsy. In any case, if we do implement the bootstrapping specified in overview.md, we should do it in different code than the Ethereum event attestations
